app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js

```diff
@@ -234,91 +195,298 @@ async function addEthereumChainHandler(
-        rpcPrefs: { blockExplorerUrl: firstValidBlockExplorerUrl },
-        chainName: _chainName,
-        rpcUrl: firstValidRPCUrl,
-        ticker,
-      },
+      type: ApprovalType.SwitchEthereumChain,
+      requestData,
     });
 
-    networkConfigurationId = await upsertNetworkConfiguration(
-      {
-        chainId: _chainId,
-        rpcPrefs: { blockExplorerUrl: firstValidBlockExplorerUrl },
-        nickname: _chainName,
-        rpcUrl: firstValidRPCUrl,
-        ticker,
-      },
-      { source: MetaMetricsNetworkEventSource.Dapp, referrer: origin },
-    );
-
-    // Once the network has been added, the requested is considered successful
+    await setActiveNetwork(networkConfigurationId);
     res.result = null;
   } catch (error) {
-    endApprovalFlow({ id: approvalFlowId });
-    return end(error);
+    // For the purposes of this method, it does not matter if the user
+    // declines to switch the selected network. However, other errors indicate
+    // that something is wrong.
+    if (error.code !== errorCodes.provider.userRejectedRequest) {
+      return end(error);
+    }
+  } finally {
+    if (approvalFlowId) {
+      endApprovalFlow({ id: approvalFlowId });
+    }
   }
+  return end();
+};
 
+const switchChain = async (
+  res,
+  end,
+  origin,
+  chainId,
+  requestData,
+  networkConfigurationId,
+  approvalFlowId,
+  {
+    getChainPermissionsFeatureFlag,
+    setActiveNetwork,
+    endApprovalFlow,
+    requestUserApproval,
+    getCaveat,
+    requestSwitchNetworkPermission,
+  },
+) => {
+  // If this network is already added but is not the currently selected network for the given origin
   // Ask the user to switch the network
-  try {
-    await requestUserApproval({
+  if (getChainPermissionsFeatureFlag()) {
+    await switchChainWithPermissions(
+      res,
+      end,
+      networkConfigurationId,
+      approvalFlowId,
+      chainId,
+      {
+        getCaveat,
+        requestSwitchNetworkPermission,
+        setActiveNetwork,
+        endApprovalFlow,
+      },
+    );
+  } else {
+    // TODO remove once CHAIN_PERMISSIONS feature flag is gone
+    await switchChainWithoutPermissions(
+      res,
+      end,
+      requestData,
+      networkConfigurationId,
+      approvalFlowId,
       origin,
-      type: ApprovalType.SwitchEthereumChain,
-      requestData: {
-        toNetworkConfiguration: {
+      { setActiveNetwork, endApprovalFlow, requestUserApproval },
+    );
+  }
+  return end();
+};
+
+async function addEthereumChainHandler(
+  req,
+  res,
+  _next,
+  end,
+  {
+    upsertNetworkConfiguration,
+    getCurrentRpcUrl,
+    findNetworkConfigurationBy,
+    setActiveNetwork,
+    requestUserApproval,
+    startApprovalFlow,
+    endApprovalFlow,
+    getCurrentChainIdForDomain,
+    getCaveat,
+    requestSwitchNetworkPermission,
+    getChainPermissionsFeatureFlag,
+  },
+) {
+  const validParams = validateParams(req.params[0], end);
+  if (!validParams) {
+    return end();
+  }
```

Comment:
This permits `end()` to be called twice, which IIRC is forbidden by `json-rpc-engine`. Better to throw inside `validateParams` and catch and `return end(error)` out here.

Reply:
Done here: https://github.com/MetaMask/metamask-extension/pull/24415/commits/4f5a98cc26676604e988a7ca9437236553f4e689